### PR TITLE
V2 Fixes Into V3

### PR DESCRIPTION
**Pull request by CryptoMaxPlanck**
**Description**
In the process of fixing up testnet, many fixes were added to #853. This PR moves those fixes into the master branch that is based on v3.

### CODE REVIEW

**Comment by CryptoMaxPlanck** [comment]
scribe and explorer tests are flaking. not entirely sure why @nautsimon @trajan0x

File: agents/agents/executor/executor.go [hunk 1]
```diff
@@ -706,21 +724,21 @@ func (e Executor) streamLogs(ctx context.Context, grpcClient pbscribe.ScribeServ
...
 	defer func() {
 		metrics.EndSpanWithErr(span, err)
 	}()
 
-	switch contractEvent.contractType {
-	case originContract:
+	//nolint:exhaustive
+	switch contractType {
+	case execTypes.OriginContract:
 		message, err := e.logToMessage(log, chainID)
```
Comment (trajan0x) [inline agents/agents/executor/executor.go:742]: There's 0 reason to ever return a pointer to an interface in go. 

e.logToMessage can return types.Message. You can still check if types.Message is nil w/ `if messsage == nil return nil`

Than you can avoid all this dereference crap below. This goes for message, snapshot, and attestation.

Pointers should only be used on structs since those can't be nil. This is one reason we prefer interfaces in those type files in general

File: agents/agents/executor/executor.go [hunk 2]
```diff
@@ -747,58 +769,71 @@ func (e Executor) processLog(parentCtx context.Context, log ethTypes.Log, chainI
...
-			if err != nil {
-				return fmt.Errorf("could not store attestation: %w", err)
+		var logHeader *ethTypes.Header
+
+	retryLoop:
+		for {
+			select {
+			case <-ctx.Done():
+				return fmt.Errorf("context canceled: %w", ctx.Err())
+			case <-time.After(timeout):
```
Comment (trajan0x) [inline agents/agents/executor/executor.go:817]: Yo, thoughts on wrapping this pattern since we do it all over the place?

File: agents/agents/executor/executor.go [hunk 3]
```diff
@@ -832,11 +867,14 @@ func (e Executor) receiveLogs(ctx context.Context, chainID uint32) error {
 //
 //nolint:gocognit,cyclop
 func (e Executor) executeExecutable(parentCtx context.Context, chainID uint32) (err error) {
```
Comment (trajan0x) [inline agents/agents/executor/executor.go:869]: [nit] no  need for parentCtx if we're not wrapping

File: agents/agents/executor/executor.go [hunk 4]
```diff
@@ -861,14 +899,18 @@ func (e Executor) executeExecutable(parentCtx context.Context, chainID uint32) (
...
-						return fmt.Errorf("could not convert message to leaf: %w", err)
+						return fmt.Errorf("could not check if message was executed: %w", err)
 					}
 
-					destinationDomain := message.DestinationDomain()
+					span.AddEvent("checked if message was executed", trace.WithAttributes(
+						attribute.Int(metrics.ChainID, int(chainID)),
+						attribute.Int(metrics.Nonce, int(message.Nonce())),
+						attribute.Bool("message_executed", messageExecuted),
+					))
```
Comment (trajan0x) [inline agents/agents/executor/executor.go:911]: thoughts on const?

File: agents/agents/executor/executor_utils.go [hunk 5]
```diff
@@ -51,50 +46,25 @@ func (e Executor) logToSnapshot(log ethTypes.Log, chainID uint32) (*types.Snapsh
 }
 
 // logType determines whether a log is a `Sent` from Origin.sol or `AttestationAccepted` from Destination.sol.
-func (e Executor) logType(log ethTypes.Log, chainID uint32) contractEventType {
-	contractEvent := contractEventType{
-		contractType: other,
-		eventType:    otherEvent,
-	}
+// TODO: Clean with switch case.
+func (e Executor) logType(log ethTypes.Log, chainID uint32) execTypes.ContractType {
```
Comment (trajan0x) [inline agents/agents/executor/executor_utils.go:50]: see my comment above about switching on .(type). Might make more sense here. Also, yes. Definitely 100% switch case.

This is a really, really important method (although more so in the notary and the guard then here).

This should be fuzzed to make sure that not only is the eventtpye correct, but it's coming from the correct contract

File: agents/agents/executor/export_for_testing.go [hunk 6]
```diff
@@ -32,20 +30,10 @@ import (
...
-	conn, err := grpc.DialContext(ctx, fmt.Sprintf("%s:%d", scribeClient.URL, scribeClient.Port), grpc.WithTransportCredentials(insecure.NewCredentials()))
-	if err != nil {
-		return nil, fmt.Errorf("could not dial grpc: %w", err)
-	}
-
-	grpcClient := pbscribe.NewScribeServiceClient(conn)
 
-	// Ensure that gRPC is up and running.
-	healthCheck, err := grpcClient.Check(ctx, &pbscribe.HealthCheckRequest{}, grpc.WaitForReady(true))
+	conn, grpcClient, err := makeScribeClient(ctx, handler, fmt.Sprintf("%s:%d", scribeClient.URL, scribeClient.Port))
```
Comment (trajan0x) [inline agents/agents/executor/export_for_testing.go:34]: unrelated, but I hate this scribeclient struct pattern. Just kinda wish it would handle more of this otel stuff for you. Not related to this pr

File: agents/agents/executor/types/db_state.go [hunk 7]
```diff
@@ -20,6 +20,8 @@ type DBState struct {
 	Proof *json.RawMessage
 	// StateIndex is the index of the state in the Snapshot.
 	StateIndex *uint32
+	// BlockNumber is the block number the state was received at on Summit.
+	BlockNumber *uint64
```
Comment (trajan0x) [inline agents/agents/executor/types/db_state.go:24]: I don't really want to address in this pr, as I think it's big enough, but I'm a really big fan of hiding this kind of db implementation stuff from the rest of the program since that can get messy

File: services/scribe/node/scribe.go [hunk 8]
```diff
@@ -109,220 +108,220 @@ func (s Scribe) Start(ctx context.Context) error {
 	return nil
 }
 
-//nolint:gocognit, cyclop
```
[the remark this answers is not in the document]
Comment (trajan0x) [reply services/scribe/node/scribe.go:112]: Hope so @nautsimon?

**Review by trajan0x — CHANGES_REQUESTED** [review]

File: agents/agents/executor/cmd/commands.go [hunk 9]
```diff
@@ -130,13 +129,12 @@ func createExecutorParameters(c *cli.Context, metrics metrics.Handler) (executor
 		ethClient := ethclient.NewClient(rpcDial)
 		clients[execClient.ChainID] = ethClient
 	} */
+
 	for _, execClient := range executorConfig.Chains {
-		rpcDial, err := rpc.DialContext(c.Context, execClient.TempRPC)
+		ethClient, err := ethergoClient.DialBackend(ctx, execClient.TempRPC, metrics, ethergoClient.Capture(true))
```
[the remark this answers is not in the document]
Comment (CryptoMaxPlanck) [reply agents/agents/executor/cmd/commands.go:134]: Just for capture? Or using ethergoClient with metrics vs just rpc dial?

File: agents/agents/executor/cmd/commands.go [hunk 10]
```diff
@@ -115,12 +112,14 @@ func createExecutorParameters(c *cli.Context, metrics metrics.Handler) (executor
...
 
-	executorDB, err = InitExecutorDB(c.Context, c.String(dbFlag.Name), c.String(pathFlag.Name), executorConfig.DBPrefix, metrics)
+	executorDB, err = InitExecutorDB(ctx, c.String(dbFlag.Name), c.String(pathFlag.Name), executorConfig.DBPrefix, metrics)
 	if err != nil {
 		return executorConfig, nil, nil, fmt.Errorf("failed to initialize database: %w", err)
 	}
 
 	clients = make(map[uint32]executor.Backend)
+
+	// TODO: The following chunk of code should be added back once we have confidence in omniRPC. For now we use tempRPC.
```
[the remark this answers is not in the document]
Comment (CryptoMaxPlanck) [reply agents/agents/executor/cmd/commands.go:122]: Ah yea, now that I am using my own omnirpc, I can use this. This was before when I was using the deployed omnirpc instead of my own.

File: agents/agents/executor/db/crosstable_test.go [hunk 11]
```diff
@@ -88,20 +89,21 @@ func (t *DBSuite) TestGetTimestampForMessage() {
...
 
-		timestamp, err = testDB.GetTimestampForMessage(t.GetTestContext(), origin, origin+1, nonceC, "")
+		retrievedTimestampC, err := testDB.GetTimestampForMessage(t.GetTestContext(), origin, origin+1, nonceC)
 		Nil(t.T(), err)
-		Equal(t.T(), uint64(1), *timestamp)
+		Equal(t.T(), uint64(1), *retrievedTimestampC)
 	})
 }
 
+// TODO: Add more edge cases.
```
[the remark this answers is not in the document]
Comment (CryptoMaxPlanck) [reply agents/agents/executor/db/crosstable_test.go:106]: For these tests, the edge cases I had in mind were just altering the origin/destination chains, but now that I think about it, I think it might over-obfuscate the main thing that these tests are trying to test.

File: agents/agents/executor/db/datastore/sql/base/crosstable.go [hunk 12]
```diff
@@ -13,34 +14,38 @@ import (
...
-		attestationsTableName = fmt.Sprintf("%s_%s", tablePrefix, attestationsTableName)
+	attestationsTableName, err := dbcommon.GetModelName(s.DB(), &Attestation{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get attestations table name: %w", err)
 	}
 
+	var timestamp uint64
+
+	// TODO: Use string formatting to make this query more legible.
 	dbTx := s.DB().WithContext(ctx).
```
[the remark this answers is not in the document]
Comment (CryptoMaxPlanck) [reply agents/agents/executor/db/datastore/sql/base/crosstable.go:31]: :p

[the remark this answers is not in the document]
Comment (CryptoMaxPlanck) [reply agents/agents/executor/db/datastore/sql/base/crosstable.go:31]: Yeah, I was going to fix it, but forgot the name of the lib you used. But yea, going to fix it now.

File: agents/agents/executor/db/datastore/sql/base/crosstable.go [hunk 13]
```diff
@@ -13,34 +14,38 @@ import (
...
 					SELECT MIN(%s) FROM (
 						(SELECT * FROM %s WHERE %s = ? AND %s >= ?) AS stateTable
 						INNER JOIN
-						(SELECT %s, %s FROM %s) AS attestationTable
+						(SELECT %s, %s FROM %s WHERE %s = ?) AS attestationTable
 						ON stateTable.%s = attestationTable.%s
 					)
-				)`,
-			DestinationTimestampFieldName, attestationsTableName, DestinationFieldName, DestinationBlockNumberFieldName,
+				) LIMIT 1`,
```
[the remark this answers is not in the document]
Comment (CryptoMaxPlanck) [reply agents/agents/executor/db/datastore/sql/base/crosstable.go:40]: yea adding order by attestation nonce descending

File: agents/agents/executor/executor.go [hunk 14]
```diff
@@ -611,36 +599,70 @@ func newTreeFromDB(ctx context.Context, chainID uint32, executorDB db.ExecutorDB
...
-	return e.streamLogs(ctx, e.grpcClient, e.grpcConn, chain.ChainID, chain.DestinationAddress, &blockNumber, contractEventType{
-		contractType: destinationContract,
-		eventType:    executedEvent,
-	})
+	for {
+		select {
+		case <-ctx.Done():
+			return false, fmt.Errorf("context canceled: %w", ctx.Err())
+		case <-time.After(timeout):
+			if b.Attempt() >= 5 {
```
[the remark this answers is not in the document]
Comment (CryptoMaxPlanck) [reply agents/agents/executor/executor.go:629]: oops. changed to `rpcRetry`

File: agents/agents/executor/executor.go (hunk 3 above)
Comment (CryptoMaxPlanck) [reply agents/agents/executor/executor.go:869]: wym? we do wrap, just later on in the function

File: agents/agents/executor/executor.go (hunk 4 above)
Comment (CryptoMaxPlanck) [reply agents/agents/executor/executor.go:911]: done

File: agents/agents/executor/types/db_state.go (hunk 7 above)
Comment (CryptoMaxPlanck) [reply agents/agents/executor/types/db_state.go:24]: Added TODO's for each db_<type>.go

File: agents/agents/executor/executor.go [hunk 15]
```diff
@@ -706,21 +724,21 @@ func (e Executor) streamLogs(ctx context.Context, grpcClient pbscribe.ScribeServ
...
 
 	defer func() {
 		metrics.EndSpanWithErr(span, err)
 	}()
 
-	switch contractEvent.contractType {
-	case originContract:
+	//nolint:exhaustive
+	switch contractType {
+	case execTypes.OriginContract:
```
[the remark this answers is not in the document]
Comment (CryptoMaxPlanck) [reply agents/agents/executor/executor.go:741]: This looks a lot more elegant. Going to implement. Thx!

File: agents/agents/executor/db/datastore/sql/sqlite/store.go [hunk 16]
```diff
@@ -39,10 +44,13 @@ func NewSqliteStore(ctx context.Context, dbPath string, handler metrics.Handler)
 
 	handler.AddGormCallbacks(gdb)
 
-	err = gdb.WithContext(ctx).AutoMigrate(base.GetAllModels()...)
-	if err != nil {
-		return nil, fmt.Errorf("could not migrate models: %w", err)
+	if !skipMigrations {
```
[the remark this answers is not in the document]
Comment (CryptoMaxPlanck) [reply agents/agents/executor/db/datastore/sql/sqlite/store.go:47]: i honestly just moved it over when we were running into MySQL problems. saw it in scribe. do you think I should remove it?

File: agents/agents/executor/executor.go [hunk 17]
```diff
@@ -559,12 +519,40 @@ func (e Executor) verifyMessageOptimisticPeriod(parentCtx context.Context, messa
...
+	var currentTime uint64
+
+retryLoop:
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf("context canceled: %w", ctx.Err())
+		case <-time.After(timeout):
+			if b.Attempt() >= rpcRetry {
+				return nil, fmt.Errorf("could not get latest header: %w", err)
```
[the remark this answers is not in the document]
Comment (CryptoMaxPlanck) [reply agents/agents/executor/executor.go:515]: what should be covered here? the failure case?

File: agents/agents/executor/cmd/commands.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (trajan0x) [reply agents/agents/executor/cmd/commands.go]: Just for capture

File: agents/agents/executor/db/datastore/sql/sqlite/store.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (trajan0x) [reply agents/agents/executor/db/datastore/sql/sqlite/store.go:47]: Nah, it's fine. Just couldn't remember

File: agents/agents/executor/executor.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (trajan0x) [reply agents/agents/executor/executor.go:515]: looks like it is now. I just wasn't seeing any coverage on `time.After` before in coveralls

File: agents/agents/executor/executor.go [hunk 18]
```diff
@@ -682,22 +701,21 @@ func (e Executor) streamLogs(ctx context.Context, grpcClient pbscribe.ScribeServ
...
-			// If we are filtering for `executed` events, we do not need to `verifyAfter`
-			// since we are backfilling.
-			if contractEvent.eventType == executedEvent {
-				e.chainExecutors[chainID].logChan <- log
-
-				continue
-			}
-			if !e.chainExecutors[chainID].lastLog.verifyAfter(*log) {
-				logger.Warnf("log is not in chronological order. last log blockNumber: %d, blockIndex: %d. this log blockNumber: %d, blockIndex: %d, txHash: %s", e.chainExecutors[chainID].lastLog.blockNumber, e.chainExecutors[chainID].lastLog.blockIndex, log.BlockNumber, log.Index, log.TxHash.String())
+			_, span := e.handler.Tracer().Start(ctx, "executor.streamLog", trace.WithAttributes(
```
[the remark this answers is not in the document]
Comment (trajan0x) [reply agents/agents/executor/executor.go:680]: Cool

File: agents/agents/executor/executor.go [hunk 19]
```diff
@@ -682,22 +676,22 @@ func (e Executor) streamLogs(ctx context.Context, grpcClient pbscribe.ScribeServ
...
-				e.chainExecutors[chainID].logChan <- log
-
-				continue
-			}
-			if !e.chainExecutors[chainID].lastLog.verifyAfter(*log) {
-				logger.Warnf("log is not in chronological order. last log blockNumber: %d, blockIndex: %d. this log blockNumber: %d, blockIndex: %d, txHash: %s", e.chainExecutors[chainID].lastLog.blockNumber, e.chainExecutors[chainID].lastLog.blockIndex, log.BlockNumber, log.Index, log.TxHash.String())
+			// We do not use a span context here because this is just meant to track transactions coming in.
+			_, span := e.handler.Tracer().Start(ctx, "executor.streamLog", trace.WithAttributes(
+				attribute.Int(metrics.ChainID, int(chainID)),
+				attribute.Int("contract", int(contractType)),
```
Comment (trajan0x) [inline agents/agents/executor/executor.go:682]: contractType should be a string

File: agents/agents/executor/executor.go [hunk 20]
```diff
@@ -706,21 +724,21 @@ func (e Executor) streamLogs(ctx context.Context, grpcClient pbscribe.ScribeServ
...
 //nolint:cyclop,gocognit
 func (e Executor) processLog(parentCtx context.Context, log ethTypes.Log, chainID uint32) (err error) {
-	contractEvent := e.logType(log, chainID)
+	contractType := e.logType(log, chainID)
 
 	ctx, span := e.handler.Tracer().Start(parentCtx, "processLog", trace.WithAttributes(
 		attribute.Int(metrics.ChainID, int(chainID)),
-		attribute.Int("contract", int(contractEvent.contractType)),
-		attribute.Int("event", int(contractEvent.eventType)),
+		attribute.Int("contract", int(contractType)),
```
[the remark this answers is not in the document]
Comment (trajan0x) [reply agents/agents/executor/executor.go]: was referring to contracttype being a string here. Added different more explicit comment on that line

File: agents/agents/executor/executor.go (hunk 3 above)
Comment (trajan0x) [reply agents/agents/executor/executor.go:820]: nah, it remains parentCtx and never gets wrapped into ctx

File: agents/agents/executor/types/db_state.go (hunk 7 above)
Comment (trajan0x) [reply agents/agents/executor/types/db_state.go:26]: if this is just a mask, let's move it to the db package. No circular dependency issues tehre

File: agents/agents/executor/cmd/commands.go [hunk 21]
```diff
@@ -115,23 +118,15 @@ func createExecutorParameters(c *cli.Context, metrics metrics.Handler) (executor
...
-		if err != nil {
-			return executorConfig, nil, nil, fmt.Errorf("failed to dial rpc: %w", err)
-		}
 
-		ethClient := ethclient.NewClient(rpcDial)
-		clients[execClient.ChainID] = ethClient
-	} */
 	for _, execClient := range executorConfig.Chains {
-		rpcDial, err := rpc.DialContext(c.Context, execClient.TempRPC)
+		rpcDial, err := rpc.DialContext(c.Context, fmt.Sprintf("%s/%d/rpc/%d", executorConfig.BaseOmnirpcURL, 1, execClient.ChainID))
```
[the remark this answers is not in the document]
Comment (CryptoMaxPlanck) [reply agents/agents/executor/cmd/commands.go]: yea sure, sounds good. lemme impl

File: agents/agents/executor/db/datastore/sql/base/crosstable.go [hunk 22]
```diff
@@ -13,34 +14,38 @@ import (
 // the same chain ID and a nonce greater than or equal to the message nonce).
 // 2. Get the minimum destination block number for all attestations that are associated to the potential snapshot roots.
 // 3. Return the timestamp of the attestation with the minimum destination block number.
-func (s Store) GetTimestampForMessage(ctx context.Context, chainID, destination, nonce uint32, tablePrefix string) (*uint64, error) {
-	var timestamp uint64
-
-	statesTableName := "states"
-	attestationsTableName := "attestations"
+func (s Store) GetTimestampForMessage(ctx context.Context, chainID, destination, nonce uint32) (*uint64, error) {
+	statesTableName, err := dbcommon.GetModelName(s.DB(), &State{})
```
[the remark this answers is not in the document]
Comment (CryptoMaxPlanck) [reply agents/agents/executor/db/datastore/sql/base/crosstable.go]: added
